lib/listen/fsm.rb

```diff
@@ -1,131 +1,128 @@
+    def transition!(new_state_name)
+      new_state_name.is_a?(Symbol) or raise ArgumentError, "state name must be a Symbol (got #{new_state_name.inspect})"
+      @mutex or raise ArgumentError, "FSM not initialized. You must call super() from initialize!"
+      @mutex.synchronize do
+        yield if block_given?
+        @state = new_state_name
+        @state_changed.broadcast
+      end
     end
 
-    protected
+    # checks for one of the given states to wait for
+    # if not already, waits for a state change (up to timeout seconds--`nil` means infinite)
+    # returns truthy iff the transition to one of the desired state has occurred
+    def wait_for_state(*wait_for_states, timeout: nil)
+      @mutex.synchronize do
+        if !wait_for_states.include?(@state)
+          @state_changed.wait(@mutex, timeout)
+        end
+        wait_for_states.include?(@state)
+      end
+    end
 
-    def validate_and_sanitize_new_state(state_name)
-      state_name = state_name.to_sym
+    protected
 
-      return if current_state_name == state_name
+    def validate_and_sanitize_new_state(new_state_name)
+      return nil if @state == new_state_name
 
-      if current_state && !current_state.valid_transition?(state_name)
+      if current_state && !current_state.valid_transition?(new_state_name)
         valid = current_state.transitions.map(&:to_s).join(', ')
-        msg = "#{self.class} can't change state from '#{@state}'"\
-          " to '#{state_name}', only to: #{valid}"
-        fail ArgumentError, msg
+        msg = "#{self.class} can't change state from '#{@state}' to '#{new_state_name}', only to: #{valid}"
+        raise ArgumentError, msg
       end
 
-      new_state = states[state_name]
-
-      unless new_state
-        return if state_name == default_state
-        fail ArgumentError, "invalid state for #{self.class}: #{state_name}"
+      unless (new_state = self.class.states[new_state_name])
+        new_state_name == self.class.start_state or raise ArgumentError, "invalid state for #{self.class}: #{new_state_name}"
       end
 
       new_state
     end
 
-    def transition_with_callbacks!(state_name)
-      transition! state_name.name
-      state_name.call(self)
-    end
-
-    def states
-      self.class.states
-    end
-
-    def default_state
-      self.class.default_state
+    def transition_with_callbacks!(new_state)
+      transition! new_state.name
+      new_state.call(self)
     end
 
     def current_state
-      states[@state]
-    end
-
-    def current_state_name
-      current_state && current_state.name || ''
+      self.class.states[@state]
     end
 
     class State
       attr_reader :name, :transitions
 
-      def initialize(name, transitions = nil, &block)
+      def initialize(name, transitions, &block)
         @name = name
         @block = block
-        @transitions = nil
-        @transitions = Array(transitions).map(&:to_sym) if transitions
+        @transitions = if transitions
```

Comment:
You don't need `if transitions`.

Reply:
It's simply not needed. `Array(nil) => []`.

Reply:
I don't follow. That `if` is there so that when `transitions` is `nil`, `@transitions` will be as well. The code below counts on that, because it uses `nil` in `@transitions` as an indicator that _all_ transitions are allowed.